File: jas/tools/svg_roundtrip.py
#!/usr/bin/env python3
"""CLI tool for cross-language commutativity testing.

Usage:
    python svg_roundtrip.py parse <file.svg>      -- parse SVG, output canonical JSON
    python svg_roundtrip.py roundtrip <file.svg>  -- parse SVG, re-serialize, output SVG
"""

import sys
import os

# Add project root to path so imports work.
sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from geometry.svg import svg_to_document, document_to_svg
from geometry.test_json import document_to_test_json


def main():
    if len(sys.argv) < 3:
        print(f"Usage: {sys.argv[0]} parse|roundtrip <file.svg>", file=sys.stderr)
        sys.exit(1)

    mode = sys.argv[1]
    filepath = sys.argv[2]

    with open(filepath) as f:
        svg = f.read()

    doc = svg_to_document(svg)

    if mode == "parse":
        print(document_to_test_json(doc), end="")
    elif mode == "roundtrip":
        print(document_to_svg(doc), end="")
    else:
        print(f"Unknown mode: {mode} (use 'parse' or 'roundtrip')", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: jas_dioxus/src/bin/svg_roundtrip.rs
/// CLI tool for cross-language commutativity testing.
///
/// Usage:
///   svg_roundtrip parse <file.svg>      -- parse SVG, output canonical JSON
///   svg_roundtrip roundtrip <file.svg>  -- parse SVG, re-serialize, output SVG

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() < 3 {
        eprintln!("Usage: {} parse|roundtrip <file.svg>", args[0]);
        std::process::exit(1);
    }

    let mode = &args[1];
    let path = &args[2];

    let svg = std::fs::read_to_string(path)
        .unwrap_or_else(|e| { eprintln!("Failed to read {}: {}", path, e); std::process::exit(1); });

    let doc = jas_dioxus::geometry::svg::svg_to_document(&svg);

    match mode.as_str() {
        "parse" => {
            print!("{}", jas_dioxus::geometry::test_json::document_to_test_json(&doc));
        }
        "roundtrip" => {
            print!("{}", jas_dioxus::geometry::svg::document_to_svg(&doc));
        }
        _ => {
            eprintln!("Unknown mode: {} (use 'parse' or 'roundtrip')", mode);
            std::process::exit(1);
        }
    }
}
